### PR TITLE
Improve cudf::cuda_error

### LINKED ISSUE #10553: [FEA] Improvements to `cudf::cuda_error`
**Is your feature request related to a problem? Please describe.**

Based on discussion in https://github.com/rapidsai/cudf/issues/10200#issuecomment-1081884311 there are a number of improvements we should make to the exceptions libcudf throws when a CUDA error occurs.

**Describe the solution you'd like**

1. Add a `cudaError_t` member to [`cudf::cuda_error`](https://github.com/rapidsai/cudf/blob/bc8f57843d2427ed07101faa86a40b162883d032/cpp/include/cudf/utilities/error.hpp#L32-L34) and corresponding `error_code()` function that returns the error code. 
2. Update `throw_cuda_error` to include the `cudaError_t` in the ctor of the `cudf::cuda_error`
3. Add a new `cudf::sticky_cuda_error` exception deriving from `cudf::cuda_error`
4. Update `throw_cuda_error` to detect and throw a sticky error, e.g.,
```
inline void throw_on_cuda_error(cudaError_t error, const char* file, unsigned int line)
   if(cudaSuccess != error){
       cudaGetLastError();                      // Clear any lingering cudart error
       auto const last = cudaGetLastError();    // Check if cudart error still exists (sticky)
       bool const is_sticky = (error == last) 
                              and (last == cudaDeviceSynchronize()); // Sync to ensure "last" wasnt an async error
       auto const msg = std::string{"CUDA error encountered at: " + std::string{file} + ":" +
                                     std::to_string(line) + ": " + std::to_string(error) + " " +
                                     cudaGetErrorName(error) + " " + cudaGetErrorString(error)};
       throw is_sticky ? sticky_cuda_error{"Sticky" + msg, error} : cudart_error{msg, error}; 
   }

#define CUDA_TRY(call)                                            \
  do {                                                            \
    cudaError_t const status = (call);                            \
    throw_on_cuda_error(status, __FILE__, __LINE__);
  } while (0)
```

### PR DESCRIPTION
**Pull request by sperlingxx**
Closes  #10553

Improves `cudf::cuda_error` in two aspects:
1. Add a cudaError_t member to `cudf::cuda_error` and corresponding error_code() function that returns the error code
2. Breaks down `cuda::cuda_error` as `sticky_cuda_error` and `cudart_error`. `sticky_cuda_error` refers to fatal error on device.

### CODE REVIEW

**Comment by sperlingxx** [comment]
I don't know how to produce a sticky error in the unit test.

**Review by karthikeyann — APPROVED** [review]

File: cpp/include/cudf/utilities/error.hpp [hunk 1]
```diff
@@ -46,7 +46,25 @@ struct logic_error : public std::logic_error {
...
+  }
+  cudaError_t error_code() { return _cudaError; }
+
+ protected:
+  cudaError_t _cudaError;
+};
+
+struct cudart_error : public cuda_error {
+  cudart_error(std::string const& message, cudaError_t const& error) : cuda_error(message, error) {}
+};
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/error.hpp:61]: I don't think we need this exception type, do we?

File: cpp/include/cudf/utilities/error.hpp [hunk 2]
```diff
@@ -46,7 +46,25 @@ struct logic_error : public std::logic_error {
...
+
+struct sticky_cuda_error : public cuda_error {
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/error.hpp:63]: I think `fatal_cuda_error` is still a more appropriate name.

File: cpp/include/cudf/utilities/error.hpp [hunk 3]
```diff
@@ -101,9 +119,16 @@ namespace detail {
 
 inline void throw_cuda_error(cudaError_t error, const char* file, unsigned int line)
 {
-  throw cudf::cuda_error(std::string{"CUDA error encountered at: " + std::string{file} + ":" +
-                                     std::to_string(line) + ": " + std::to_string(error) + " " +
-                                     cudaGetErrorName(error) + " " + cudaGetErrorString(error)});
+  cudaGetLastError();
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/error.hpp:122]: Let's add the doc I wrote about why this algorithm is doing what it's doing.

File: cpp/include/cudf/utilities/error.hpp (hunk 1 above)
Comment (sperlingxx) [reply cpp/include/cudf/utilities/error.hpp:61]: So, we parse the message to get the exact cuda error ?

File: cpp/include/cudf/utilities/error.hpp (hunk 2 above)
Comment (sperlingxx) [reply cpp/include/cudf/utilities/error.hpp]: Changed.

File: cpp/include/cudf/utilities/error.hpp (hunk 3 above)
Comment (sperlingxx) [reply cpp/include/cudf/utilities/error.hpp:121]: Doc added

File: cpp/include/cudf/utilities/error.hpp (hunk 1 above)
Comment (sperlingxx) [reply cpp/include/cudf/utilities/error.hpp]: Hi @jrhemstad, I am not sure whether you refer to simplify the constructor with `using` or remove the member `error` .  I think the cuda_error parameter may be helpful for caller (like cuDF JNI) to demonstrate the error message.

File: cpp/include/cudf/utilities/error.hpp [hunk 4]
```diff
@@ -119,13 +118,17 @@ namespace detail {
...
   cudaGetLastError();
   auto const last = cudaGetLastError();
   auto const msg  = std::string{"CUDA error encountered at: " + std::string{file} + ":" +
                                std::to_string(line) + ": " + std::to_string(error) + " " +
                                cudaGetErrorName(error) + " " + cudaGetErrorString(error)};
+  // Calls cudaDeviceSynchronize to make sure that there is no other asynchronize error occurs
+  // between two calls.
   if (error == last && last == cudaDeviceSynchronize()) {
-    throw sticky_cuda_error{"Sticky " + msg, error};
+    throw fatal_cuda_error{"Sticky " + msg, error};
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/error.hpp:131]:
```suggestion
    throw fatal_cuda_error{"Fatal " + msg, error};
```

Comment (sperlingxx) [reply cpp/include/cudf/utilities/error.hpp]: Fixed.

File: cpp/include/cudf/utilities/error.hpp (hunk 1 above)
Comment (sperlingxx) [reply cpp/include/cudf/utilities/error.hpp]: Yeah. I removed `cudart_error`

File: cpp/include/cudf/utilities/error.hpp [hunk 5]
```diff
@@ -115,13 +139,10 @@ inline void throw_cuda_error(cudaError_t error, const char* file, unsigned int l
...
-  do {                                                            \
-    cudaError_t const status = (call);                            \
-    if (cudaSuccess != status) {                                  \
-      cudaGetLastError();                                         \
-      cudf::detail::throw_cuda_error(status, __FILE__, __LINE__); \
-    }                                                             \
+#define CUDF_CUDA_TRY(call)                                                                    \
+  do {                                                                                         \
+    cudaError_t const status = (call);                                                         \
+    if (cudaSuccess != status) { cudf::detail::throw_cuda_error(status, __FILE__, __LINE__); } \
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/error.hpp:145]: The semicolon after the `while(0)` should be removed to ensure all uses of the `CUDF_CUDA_TRY` macro are terminated with a semi-colon.

**Comment by jrhemstad** [comment]
> I don't know how to produce a sticky error in the unit test.

Hm, I think calling `assert` in device code should generate a sticky error.

**Comment by sperlingxx** [comment]
@gpucibot merge

**Pull request merged**